src/charm.py

```diff
@@ -210,21 +253,100 @@ def _configure(self, event: EventBase) -> None:
+                refresh=True
+            )
+        except SecretNotFoundError:
+            logger.warning("No active CA certificate found to move to expiring")
+            return
+        self._set_juju_secret(
+            label=EXPIRING_CA_CERTIFICATES_SECRET_LABEL,
+            content=current_active_ca_cert_secret_content,
+            expire=timedelta(days=self._config_certificate_validity),
+        )
+
+    def _set_juju_secret(self, label: str, content: dict[str, str], expire: timedelta) -> None:
+        """Set a juju secret."""
+        try:
+            secret = self.model.get_secret(label=label)
+            secret.set_content(content=content)
+            secret.set_info(expire=expire)
+        except SecretNotFoundError:
+            self.app.add_secret(content=content, label=label, expire=expire)
+
     def _root_certificate_matches_config(self) -> bool:
         """Return whether the stored root certificate matches with the config."""
         if not self._config_ca_common_name:
             raise ValueError("CA common name should not be empty")
         ca_certificate_secret = self.model.get_secret(label=CA_CERTIFICATES_SECRET_LABEL)
         ca_certificate_secret_content = ca_certificate_secret.get_content(refresh=True)
         ca = ca_certificate_secret_content["ca-certificate"]
-        return self._config_ca_common_name == Certificate.from_string(ca).common_name
+        certificate = Certificate.from_string(ca)
+        certificate_validity = (
+            (certificate.expiry_time - certificate.validity_start_time).days
+            if certificate.validity_start_time and certificate.expiry_time
+            else 0
+        )
+        return (
+            self._config_ca_common_name == certificate.common_name
+            and self._config_ca_organization == certificate.organization
+            and self._config_ca_organizational_unit == certificate.organizational_unit
+            and self._config_ca_email_address == certificate.email_address
+            and self._config_ca_country_name == certificate.country_name
+            and self._config_ca_state_or_province_name == certificate.state_or_province_name
+            and self._config_ca_locality_name == certificate.locality_name
+            and self._config_root_ca_certificate_validity == certificate_validity
+        )
+
+    def _revoke_certificates(self):
+        """Revoke all certificates.
+
+        Revokes both root and issued certificates.
+        """
+        try:
+            expiring_ca_secret = self.model.get_secret(label=EXPIRING_CA_CERTIFICATES_SECRET_LABEL)
+            expiring_ca_secret.remove_all_revisions()
+        except SecretNotFoundError:
+            pass
+        try:
+            active_ca_secret = self.model.get_secret(label=CA_CERTIFICATES_SECRET_LABEL)
+            active_ca_secret.remove_all_revisions()
+        except SecretNotFoundError:
+            pass
+        self.tls_certificates.revoke_all_certificates()
```

Comment:
I suggest not implementing revoking part of this PR and have this done properly on its own. If we really want to do revocation we would need a CRL and we'd need to share it with charms. My suggestion is to leave this aside, spec it and implement it properly.

Reply:
As discussed, we will keep this "revocation" mechanism as it triggers the requirers to re-generate CSRs and get new certificates.
This should be done on private-key rotation and on CA config changes and won't happen on CA renewals.